### PR TITLE
pubsub: factor out DefaultAckBatcher func

### PR DESCRIPTION
**Pull request by ijt**
This is a step towards fixing #975.

### CODE REVIEW

File: internal/pubsub/pubsub.go [hunk 1]
```diff
@@ -257,18 +257,34 @@ func (s *Subscription) As(i interface{}) bool {
...
+
+// MakeAckBatchHandler makes a handler func suitable for constructing a Batcher
+// that can be used as an argument to NewSubscription. This handler wraps
+// d.SendAcks with retry logic. MakeAckBatchHandler is for use by provider
+// implementations.
+func MakeAckBatchHandler(ctx context.Context, d driver.Subscription) func(interface{}) error {
+	return func(items interface{}) error {
+		ids := items.([]driver.AckID)
+		// TODO: Consider providing a way to stop this call. See #766.
+		return retry.Call(ctx, gax.Backoff{}, d.IsRetryable, func() error {
```
Comment (vangent) [inline internal/pubsub/pubsub.go]: Do you want the retry behavior for tests?

File: internal/pubsub/pubsub.go [hunk 2]
```diff
@@ -257,18 +257,34 @@ func (s *Subscription) As(i interface{}) bool {
...
+	handler := MakeAckBatchHandler(callCtx, d)
+	ackBatcher := batcher.New(reflect.TypeOf([]driver.AckID{}).Elem(), maxHandlers, handler)
+	return NewSubscriptionWithBatcher(d, ackBatcher, cancel)
+}
+
+// NewSubscriptionWithBatcher creates a Subscription from a driver.Subscription
+// and a batcher that sends batches of acks to the provider. Typically the
+// batcher is constructed using the result of MakeAckBatchHandler.
+// NewSubscriptionWithBatcher is for use by provider implementations.
+func NewSubscriptionWithBatcher(d driver.Subscription, ab driver.Batcher, cancel func()) *Subscription {
```
Comment (vangent) [inline internal/pubsub/pubsub.go]: Another way to do this (not sure if it's better, but consider it):

```
func NewSubscription(d driver.Subscription, makeBatcher func(ctx context.Context, handler func(interface{}) error)) {
  callCtx, cancel := context.WithCancel(context.Background())
  handler := /* handler code inline here */
  batcher := makeBatcher(callCtx, handler)
  return &Subscription{...}
}
```

and then have a `DefaultBatcher` function that provider implementations can pass to as `makeBatcher`. Tests would pass something like `makeTestBatcher`.

Comment (vangent) [reply internal/pubsub/pubsub.go]: This avoids the awkwardness of exposing the `MakeAckBatchHandler` function, but requiring the caller to pass in a `driver.Batcher`, so they have to connect those dots.

File: internal/pubsub/pubsub.go [hunk 3]
```diff
@@ -257,18 +257,34 @@ func (s *Subscription) As(i interface{}) bool {
 // It is for use by provider implementations.
 func NewSubscription(d driver.Subscription) *Subscription {
```
[the remark this answers is not in the document]
Comment (ijt) [reply internal/pubsub/pubsub.go]: Added `DefaultAckBatcher`.

File: internal/pubsub/pubsub.go (hunk 1 above)
Comment (ijt) [reply internal/pubsub/pubsub.go]: It doesn't seem to be hurting anything.

File: internal/pubsub/pubsub.go (hunk 2 above)
Comment (ijt) [reply internal/pubsub/pubsub.go]: Done.

**Comment by ijt** [comment]
PTAL

File: pubsub/pubsub.go [hunk 4]
```diff
@@ -280,18 +280,30 @@ func (s *Subscription) As(i interface{}) bool {
...
+	ctx, cancel := context.WithCancel(context.Background())
 	s := &Subscription{
 		driver: d,
 		cancel: cancel,
 	}
+	s.ackBatcher = newAckBatcher(ctx, s)
+	return s
+}
+
+type BatcherMaker func(context.Context, *Subscription) driver.Batcher
```
[the remark this answers is not in the document]
Comment (ijt) [reply pubsub/pubsub.go]: `DefaultAckBatcher` also uses `s.mu` and `s.err`, which is why I changed it to take a `*Subscription`. The diff hides this by default; you have to expand something to see it.

[the remark this answers is not in the document]
Comment (vangent) [reply pubsub/pubsub.go]: Ah, sorry, I missed that.

**Comment by jba** [comment]
A simpler idea would be a variable `driver.AckBatcher`, which would be set by drivertest. Then nothing in the concrete API would look different.

**Comment by jba** [comment]
If you don't like global variables, then the driver interface could have an AckBatcher method.

Another idea: the drivertest package could have the global, and the pubsub package could depend on drivertest. Even I don't like that one, though.

**Comment by jba** [comment]
Just looked at pubsub.go carefully. We definitely don't want two new exported symbols, BatcherMaker and DefaultAckBatcher, just to support tests.

How about a `driver.SetAckBatcher` function. If you look at what we did with `NewPageInfo` in https://godoc.org/google.golang.org/api/iterator, using a variable assigned to an unexported function, we completely hide the function signature.

**Comment by ijt** [comment]
I just had a look and it seems that it's not really feasible to add a `driver.Subscription.AckBatcher` method since the ack batcher handler has to know about the `pubsub.Subscription` to set its `err` field in case of an unrecoverable error. 

I removed the `BatcherMaker` type, so now there's less API exposed just for tests. PTAL.

**Comment by jba** [comment]
Why does DefaultAckBatcher need to be exported? Couldn't non-test code pass nil?

**Comment by ijt** [comment]
I think that's possible. Trying it.

**Comment by ijt** [comment]
Done.

**Comment by ijt** [comment]
Done. PTAL.
